pkg/services/sqlstore/datasource.go

```diff
@@ -101,6 +102,14 @@ func AddDataSource(cmd *models.AddDataSourceCommand) error {
 			cmd.JsonData = simplejson.New()
 		}
 
+		if cmd.Uid == "" {
+			uid, err := generateNewDatasourceUid(sess, cmd.OrgId)
+			if err != nil {
+				return errutil.Wrapf(err, "Failed to generate Uid for Datasource %s", cmd.Name)
+			}
+			cmd.Uid = uid
+		}
```

Comment:
If uid provided and uid already exists on another data source should return error similar to `models.ErrDataSourceNameExists` above.

Reply:
Added mapping from unique constraint error later on.